mlos_core/mlos_core/data_classes.py

```diff
@@ -0,0 +1,350 @@
+            config = pd.concat([obs._config.to_frame().T for obs in observations])
+            score = pd.concat([obs._score.to_frame().T for obs in observations])
+
+            if sum([obs._context is None for obs in observations]) == 0:
+                context = pd.concat(
+                    [obs._context.to_frame().T for obs in observations]  # type: ignore
+                )
+            else:
+                context = None
+            if sum([obs._metadata is None for obs in observations]) == 0:
+                metadata = pd.concat(
+                    [obs._metadata.to_frame().T for obs in observations]  # type: ignore
+                )
+            else:
+                metadata = None
+        assert len(config.index) == len(score.index), "config and score must have the same length"
+        if context is not None:
+            l1 = len(config.index)
+            l2 = len(context.index)
+            assert l1 == l2, "config and context must have the same length"
+        if metadata is not None:
+            assert len(config.index) == len(
+                metadata.index
+            ), "config and metadata must have the same length"
+        self._config = config.reset_index(drop=True)
+        self._score = score.reset_index(drop=True)
+        self._context = None if context is None else context.reset_index(drop=True)
+        self._metadata = None if metadata is None else metadata.reset_index(drop=True)
+
+    def filter_by_index(self, index: pd.Index) -> Observations:
+        """
+        Filters the observation by the given indices.
+
+        Parameters
+        ----------
+        index : pd.Index
+            The indices to filter by.
+
+        Returns
+        -------
+        Observation
+            The filtered observation.
+        """
+        return Observations(
+            config=self._config.loc[index].copy(),
+            score=self._score.loc[index].copy(),
+            context=None if self._context is None else self._context.loc[index].copy(),
+            metadata=None if self._metadata is None else self._metadata.loc[index].copy(),
+        )
+
+    def append(self, other: Observation) -> None:
+        """
+        Appends the given observation to this observation.
+
+        Parameters
+        ----------
+        other : Observation
+            The observation to append.
+        """
+        config = other._config.to_frame().T
+        score = other._score.to_frame().T
+        context = None if other._context is None else other._context.to_frame().T
+        metadata = None if other._metadata is None else other._metadata.to_frame().T
+        if len(self._config.index) == 0:
+            self._config = config
+            self._score = score
+            self._context = context
+            self._metadata = metadata
+            assert set(self._config.index) == set(
+                self._score.index
+            ), "config and score must have the same index"
```

Comment:
complete nit, but can you put the string in parens so it breaks on that instead of the equality expression?

Reply:
black formats it back